ml-agents/mlagents/trainers/ghost/trainer.py

```diff
@@ -0,0 +1,256 @@
+
+        self_play_parameters = trainer_parameters["ghost"]
+        self.window = self_play_parameters["window"]
+        self.current_prob = self_play_parameters["current_prob"]
+        self.steps_between_snapshots = self_play_parameters["snapshot_per"]
+
+        self.policies: Dict[str, TFPolicy] = {}
+        self.policy_snapshots: List[Any] = [None for _ in range(self.window)]
+        self.snapshot_counter: int = 0
+        self.learning_behavior_name: str = None
+        self.current_policy_snapshot = None
+        self.last_step = 0
+
+        self.initial_elo: float = 1200.0
+        self.current_elo: float = self.initial_elo
+        self.policy_elos: List[float] = [self.initial_elo for _ in range(self.window)]
+        self.current_opponent: int = 0
+
+    @property
+    def get_step(self) -> int:
+        """
+        Returns the number of steps the trainer has performed
+        :return: the step count of the trainer
+        """
+        return self.trainer.get_step
+
+    def _write_summary(self, step: int) -> None:
+        """
+        Saves training statistics to Tensorboard.
+        """
+        opponents = np.array(self.policy_elos)
+        LOGGER.info(
+            " ELO: {}\n"
+            "Mean Opponent ELO: {}"
+            " Std Opponent ELO: {}".format(
+                self.current_elo, opponents.mean(), opponents.std()
+            )
+        )
+        self.stats_reporter.add_stat("ELO", self.current_elo)
+
+    def _process_trajectory(self, trajectory: Trajectory) -> None:
+        if (
+            trajectory.done_reached
+            and not trajectory.max_step_reached
+            and self.current_opponent > -1
+        ):
+            result = "win"
+            final_reward = trajectory.steps[-1].reward
+            if final_reward < 0:
+                result = "loss"
+            change = compute_elo_rating_changes(
+                self.current_elo, self.policy_elos[self.current_opponent], result
+            )
+            self.current_elo += change
+            self.policy_elos[self.current_opponent] -= change
+
+    def _is_ready_update(self) -> bool:
+        pass
+
+    def _update_policy(self) -> None:
+        pass
+
+    def advance(self) -> None:
+        """
+        Steps the trainer, taking in trajectories and updates if ready.
+        """
+        for traj_queue, internal_traj_queue in zip(
+            self.trajectory_queues, self.internal_trajectory_queues
+        ):
+            try:
+                t = traj_queue.get_nowait()
+                # adds to wrapped trainers queue
+                internal_traj_queue.put(t)
+                self._process_trajectory(t)
+            except AgentManagerQueue.Empty:
+                pass
+
+        self.next_update_step = self.trainer.next_update_step
+        self.trainer.advance()
+        self._maybe_write_summary(self.get_step)
+
+        for q, internal_q in zip(self.policy_queues, self.internal_policy_queues):
```

Comment:
Is it possible for these two lists to be different lengths? It seems like you only create the internal policy queue if the behavior name matches right? So not all policy_queues will have a corresponding internal_policy_queue?

Reply:
good catch

Reply:
@ervteng are you happy with the fix?